src/Stream/BufferedStream.php

```diff
@@ -0,0 +1,264 @@
+<?php
+
+namespace Http\Message\Stream;
+
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * A buffered stream allow to buffer over an existing.
+ *
+ * You should use this decorator when you want to seek over a not seekable stream.
+ * This stream is however read-only.
+ */
+class BufferedStream implements StreamInterface
+{
+    /** @var resource The buffered resource used to seek previous data */
+    private $resource;
+
+    /** @var int size of the stream if available */
+    private $size;
+
+    /** @var StreamInterface The underlying stream decorated by this class */
+    private $stream;
+
+    /** @var int How many bytes were written */
+    private $writed = 0;
+
+    public function __construct(StreamInterface $stream, $useFileBuffer = true, $memoryBuffer = 2097152)
+    {
+        $this->stream = $stream;
+        $this->size = $stream->getSize();
+
+        if ($useFileBuffer) {
+            $this->resource = fopen('php://temp/maxmemory:'.$memoryBuffer, 'rw+');
+        } else {
+            $this->resource = fopen('php://memory', 'rw+');
+        }
+
+        if (false === $this->resource) {
+            throw new \RuntimeException('Cannot create a resource over temp or memory implementation');
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        try {
+            $this->rewind();
+
+            return $this->getContents();
+        } catch (\Throwable $throwable) {
+            return '';
+        } catch (\Exception $exception) {
+            return '';
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function close()
+    {
+        if (null === $this->resource) {
+            throw new \RuntimeException('Cannot close on a detached stream');
+        }
+
+        $this->stream->close();
+        fclose($this->resource);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function detach()
+    {
+        if (null === $this->resource) {
+            return;
+        }
+
+        // Force reading the remaining data of the stream
+        $this->getContents();
+
+        $resource = $this->resource;
+        $this->stream = null;
+        $this->resource = null;
+
+        return $resource;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSize()
+    {
+        if (null === $this->size && $this->stream->eof()) {
+            if (null === $this->resource) {
```

Comment:
should this check not be outside the other if? or do we return a size if we have one even if resource has since been set to null?